in-progress/8757-fees/design.md

```diff
@@ -0,0 +1,189 @@
+- `target_mana_per_block` - the amount of mana that an "average" block is expected to consume
+- `blobs_per_block` - the number of blobs per block that are compensated for in the base fee
+- `l2_slots_per_l2_epoch` - the number of L2 slots in each L2 epoch
+- `l1_gas_per_block_proposed` - the amount of L1 gas required to propose an L2 block on L1
+- `l1_gas_per_epoch_verified` - the amount of L1 gas required to verify an L2 epoch on L1
+- `minimum_proving_cost_per_mana` - the minimum cost in wei for proving a unit of mana
+- `maximum_proving_cost_multiplier_per_block` - the maximum percentage increase in the cost of proving a unit of mana per block
+- `minimum_fee_asset_price` - the minimum price in wei of the fee asset (e.g. TST)
+- `maximum_fee_asset_price_multiplier_per_block` - the maximum percentage increase in the price of the fee asset per block
+- `underlying_base_fee_oracle_update_interval` - the minimum number of slots between updates to the underlying base fee oracle
+- `congestion_factor_multiplier` - the constant factor to multiply the congestion factor by
+- `maximum_epoch_proof_quote_fee` - the maximum fee a prover can submit for an epoch proof quote
+
+## Oracles
+
+### `fee_asset_per_wei`
+
+This will be stored on the rollup contract. On proposing a block, proposers will be able to adjust the rate by up to `maximum_fee_asset_price_multiplier_per_block`.
+
+Has a minimum value of `minimum_fee_asset_price`.
+
+### `proving_cost_per_mana`
+
+This will be stored on the rollup contract. On proposing a block, proposers will be able to adjust the rate by up to `maximum_proving_cost_multiplier_per_block`.
+
+Has a minimum value of `minimum_proving_cost_per_mana`.
+
+### `wei_per_l1_gas` and `wei_per_l1_blob_gas`
+
+The rollup contract will maintain an `wei_per_l1_gas` oracle. 
+
+Same for `wei_per_l1_blob_gas`.
+
+They will be updated up to every `underlying_base_fee_oracle_update_interval` slots.
+
+
+## Getting the cost of a block in wei/mana
+
+When a proposer is building a block, it calculates the total L1 cost of the block in wei by summing:
+- `l1_gas_per_block_proposed` * `wei_per_l1_gas`
+- `blobs_per_block` * `GAS_PER_BLOB` * `wei_per_l1_blob_gas`
+- `l1_gas_per_epoch_verified` * `wei_per_l1_gas` / `l2_slots_per_l2_epoch`
+
+The cost of the block in wei/mana is then the sum of the following:
+- the total L1 cost in wei divided by `target_mana_per_block`
+- `proving_cost_per_mana`
+
+## Getting the Base Fee in wei/mana
+
+After simulating a block, the mana used by each transaction is known.
+
+A proposer can compute a block's `total_mana_used` by summing the `mana_used` fields of all transactions in the block.
+
+This `total_mana_used` will be part of the L2 block header.
+
+The rollup contract will maintain a `excess_mana` variable for the proven and pending chains.
+
+The `excess_mana` is computed from the previous block as the difference between `total_mana_used` and `target_mana_per_block`.
```

Comment:
Not fully. It is not just the diff and the diff of the last it is an accumulated excess. If not accumulated we would max out much lower.

```python
spent = self.headers[-1].mana_used
excess = self.headers[-1].excess_mana
if excess + spent < self.mana_target:
    return 0
return excess + spent - self.mana_target
```

or I guess you can also write it as this
```python
diff = spent - self.mana_target
excess = max(0, excess + diff)
```